src/ggml-cuda.h

```diff
@@ -41,6 +43,10 @@ GGML_API bool   ggml_cuda_compute_forward(struct ggml_compute_params * params, s
 GGML_API int    ggml_cuda_get_device_count(void);
 GGML_API void   ggml_cuda_get_device_description(int device, char * description, size_t description_size);
 
+// backend API
+GGML_API ggml_backend_t ggml_backend_cuda_init(void); // TODO: take a list of devices to use
```

Comment:
Hey, it will be nice if besides devices, one could provide stream handles and cublas handles here. This will allow plugging ggml in existing code.

I currently have hacked ggml_init_cublas as it is in master to allow exactly this. Thus I created a pytorch plugin which uses the cuda instance from pytorch and now I directly assign pytorch tensor pointers to ggml tensors.

I was planning to make my hack more presentable, but this PR is basically the way to go.

Reply:
In the future, support for the old API will be removed, and then all the globals will be moved into the backend context. At that point we could consider how to add this functionality, but I think it is too early for that.

Reply:
What's wrong with using the globals? 

If one wants to add a plugin, it is very unlikely for them to need multiple backend instances. And even if they do, this can be marked as "not supported yet".

If one wants to use the old behavior, they won't provide handles and it will be just the same.

No existing functionality is going to break if this feature is added with the PR. On the contrary: plugins like mine will be possible with vanilla ggml

Reply:
There are a few problems with this:
- The backends interface is still not finished, many things are going to change in the next weeks
- I don't really like the idea exposing internals of the CUDA backend, especially at a point where things are going to change significantly
- This looks like an extremely niche application, so I am not convinced that it needs to be merged. The patch can be maintained in the relevant repository.
- I do not understand what are your requirements, so I cannot really do this myself in the first place

Reply:
I wouldn't say that it's a niche application. Consider an existing inference codebase which doesn't use ggml. Be it pytorch-based, pure cuda, or tensorflow. If one wants to migrate it to ggml for some reason, what can they do? 

They can completely rewrite the inference in ggml which is quite error prone. Or they can implement parts of it in ggml. Leaves of the computation graph and leave the rest as is. Gradually moving to entirely ggml-based inference while having a working version at each step.

This is exactly my use case. I'm moving a pytorch-based application to ggml.

I think this can be quite beneficial for the library and people who're looking into evaluate it, without having to go all the way.

Reply:
Would a function to change the main stream work? That would allow you to queue the ggml-cuda computations on the pytorch stream, which I may improve performance due to reduced synchronizaiton. Is there any reason to re-use the pytorch cuBLAS handle? Does each cuBLAS handle use a large amount of GPU resources?